exchanges/huobi/huobi_websocket.go

```diff
@@ -75,427 +73,240 @@ var subscriptionNames = map[string]string{
+			return fmt.Errorf("%w: %s:%s", stream.ErrNoMessageListener, action, ch)
 		}
-	default:
-		h.Websocket.DataHandler <- stream.UnhandledMessageWarning{
-			Message: h.Name + stream.UnhandledMessage + string(respRaw),
-		}
-		return nil
 	}
 	return nil
 }
 
-func (h *HUOBI) sendPingResponse(pong int64) {
-	err := h.Websocket.Conn.SendJSONMessage(context.TODO(), request.Unset, WsPong{Pong: pong})
-	if err != nil {
-		log.Errorln(log.ExchangeSys, err)
-	}
+func (h *HUOBI) wsHandleChannelMsgs(s *subscription.Subscription, respRaw []byte) error {
+	switch s.Channel {
+	case subscription.TickerChannel:
+		return h.wsHandleTickerMsg(s, respRaw)
+	case subscription.OrderbookChannel:
+		return h.wsHandleOrderbookMsg(s, respRaw)
+	case subscription.CandlesChannel:
+		return h.wsHandleCandleMsg(s, respRaw)
+	case subscription.AllTradesChannel:
+		return h.wsHandleAllTradesMsg(s, respRaw)
+	case subscription.MyAccountChannel:
+		return h.wsHandleMyAccountMsg(respRaw)
+	case subscription.MyOrdersChannel:
+		return h.wsHandleMyOrdersMsg(s, respRaw)
+	case subscription.MyTradesChannel:
+		return h.wsHandleMyTradesMsg(s, respRaw)
+	}
+	return fmt.Errorf("%w: %s", common.ErrNotYetImplemented, s.Channel)
 }
 
-// WsProcessOrderbook processes new orderbook data
-func (h *HUOBI) WsProcessOrderbook(update *WsDepth, symbol string) error {
-	pairs, err := h.GetEnabledPairs(asset.Spot)
-	if err != nil {
+func (h *HUOBI) wsHandleCandleMsg(s *subscription.Subscription, respRaw []byte) error {
+	if len(s.Pairs) != 1 {
+		return subscription.ErrNotSinglePair
+	}
+	var c WsKline
+	if err := json.Unmarshal(respRaw, &c); err != nil {
 		return err
 	}
+	h.Websocket.DataHandler <- stream.KlineData{
+		Timestamp:  time.UnixMilli(c.Timestamp),
+		Exchange:   h.Name,
+		AssetType:  s.Asset,
+		Pair:       s.Pairs[0],
+		OpenPrice:  c.Tick.Open,
+		ClosePrice: c.Tick.Close,
+		HighPrice:  c.Tick.High,
+		LowPrice:   c.Tick.Low,
+		Volume:     c.Tick.Volume,
+		Interval:   s.Interval.String(),
+	}
+	return nil
+}
 
-	format, err := h.GetPairFormat(asset.Spot, true)
-	if err != nil {
+func (h *HUOBI) wsHandleAllTradesMsg(s *subscription.Subscription, respRaw []byte) error {
+	if !h.IsSaveTradeDataEnabled() {
+		return nil
+	}
+	if len(s.Pairs) != 1 {
+		return subscription.ErrNotSinglePair
+	}
+	var t WsTrade
+	if err := json.Unmarshal(respRaw, &t); err != nil {
 		return err
 	}
+	trades := make([]trade.Data, 0, len(t.Tick.Data))
+	for i := range t.Tick.Data {
+		side := order.Buy
+		if t.Tick.Data[i].Direction != "buy" {
+			side = order.Sell
+		}
+		trades = append(trades, trade.Data{
+			Exchange:     h.Name,
+			AssetType:    s.Asset,
+			CurrencyPair: s.Pairs[0],
+			Timestamp:    time.UnixMilli(t.Tick.Data[i].Timestamp),
+			Amount:       t.Tick.Data[i].Amount,
+			Price:        t.Tick.Data[i].Price,
+			Side:         side,
+			TID:          strconv.FormatFloat(t.Tick.Data[i].TradeID, 'f', -1, 64),
+		})
+	}
+	return trade.AddTradesToBuffer(h.Name, trades...)
```

Comment:
Future suggestion: Maybe we can have this and above processing in engine websocket routine handler, and return this. There is a bit of toil going on here for just trades.

Reply:
Response from Future:
Yeah. I wasn't happy with what I saw in trades in general and have an open item to review it both in Huobi and generally.